doc/designs/quic-design/dgram-api.md

```diff
@@ -0,0 +1,487 @@
+    seem like a common use case.)
+
+  - Flexible support for single-copy (zero-copy).
+
+Cons of this approach:
+
+  - Very different way of doing reads/writes might be strange to existing
+    applications. *But* the primary consumer of this new API will be our own
+    QUIC implementation so probably not a big deal. We can always support
+    `BIO_read`/`BIO_write` as a less efficient fallback for existing third party
+    users of BIO_dgram.
+
+### Compatibility interop
+
+Suppose the following sequence happens:
+
+1. BIO_read (legacy call path)
+2. BIO_read_dequeue (`recvmmsg` based call path with callback-allocated buffer)
+3. BIO_read (legacy call path)
+
+For (1) we have two options
+
+a. Use `recvmmsg` and add the received datagrams to an RX queue just as for the
+   `BIO_read_dequeue` path. We use an OpenSSL-provided default allocator
+   (`OPENSSL_malloc`) and flag these datagrams as needing to be freed by OpenSSL,
+   not the application.
+
+   When the application calls `BIO_read`, a copy is performed and the internal
+   buffer is freed.
+
+b. Use `recvfrom` directly. This means we have a `recvmmsg` path and a
+   `recvfrom` path depending on what API is being used.
+
+   The disadvantage of (a) is it yields an extra copy relative to what we have now,
+   whereas with (b) the buffer passed to `BIO_read` gets passed through to the
+   syscall and we do not have to copy anything.
+
+   Since we will probably need to support platforms without
+   `sendmmsg`/`recvmmsg` support anyway, (b) seems like the better option.
+
+For (2) the new API is used. Since the previous call to BIO_read is essentially
+“stateless” (it's just a simple call to `recvfrom`, and doesn't require mutation
+of any internal BIO state other than maybe the last datagram source/destination
+address fields), BIO_dgram can go ahead and start using the `recvmmsg` code
+path. Since the RX queue will obviously be empty at this point, it is
+initialised and filled using `recvmmsg`, then one datagram is popped from it.
+
+For (3) we have a legacy `BIO_read` but we have several datagrams still in the
+RX queue. In this case we do have to copy - we have no choice. However this only
+happens in circumstances where a user of BIO_dgram alternates between old and
+new APIs, which should be very unusual.
+
+Subsequently for (3) we have to free the buffer using the free callback. This is
+an unusual case where BIO_dgram is responsible for freeing read buffers and not
+the application (the only other case being premature destruction, see below).
+But since this seems a very strange API usage pattern, we may just want to fail
+in this case.
+
+Probably not worth supporting this. So we can have the following rule:
+
+- After the first call to `BIO_read_dequeue` is made on a BIO_dgram, all
+  subsequent calls to ordinary `BIO_read` will fail.
+
+Of course, all of the above applies analogously to the TX side.
+
+#### BIO_dgram_mem
```

Comment:
This section seems to be a subsection of "Compatibility interop" - which is odd. It seems applicable to the selected design.